### PR TITLE
ImproperSerializeField analyzer

### LINKED ISSUE #14: Warn about improper usage of SerializeField
#### Problem statement
Because the `SerializeField` attribute does not have an `AttributeUsage` attribute limiting its usage to fields, people sometimes mistakenly use the `SerializeField` on properties. We should have an analyzer warning about it.

Also, we've seen people using `SerializeField` on a public field, which is not necessary either.

#### Proposed solution
In both cases, we can safely offer to remove the attribute.

### PR DESCRIPTION
**Pull request by AnBucyk**
Fixes #14

#### Checklist

- [X] I have read the [Contribution Guide](/CONTRIBUTING.md) ;
- [X] There is an approved issue describing the change when contributing a new analyzer or suppressor ;
- [X] I have added tests that prove my fix is effective or that my feature works ;
- [X] I have added necessary documentation (if appropriate) ;

#### Short description of what this resolves:

This PR adds an analyzer that warns about improper uses of the `SerializeField` attribute. It also provides a code fix suggestion to automatically remove improper uses of the attribute from the code.

#### Notes

I'm not sure if my solution for checking fields is correct. Field declarations can include multiple symbols. In this case, I iterate through the symbols, and any that meet the requirements results in the warning and code fix suggestion. Could use another set of eyes on it - does this look correct to you?

### CODE REVIEW

File: src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+		{
+			var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+			var attributes = new SyntaxList<AttributeListSyntax>();
+
+			foreach (var attributeList in declaration.AttributeLists)
+			{
+				var nodes = attributeList
+					.Attributes
+					.Where(a => a.Name.ToString() == "SerializeField");
```
Comment (jbevain) [inline src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs]: Let's avoid comparing names. We want to make sure that we're removing the proper attribute so we need to work at the symbol level and enforce that it's actually a UnityEngine.SerializeField, like you do in the diagnostic above.

File: src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+			{
+				var symbol = context.SemanticModel.GetDeclaredSymbol(variable);
+				if (symbol != null)
+					symbols.Add(symbol);
+			}
+
+			if (symbols.FirstOrDefault() == null)
+				return;
+
+			if (!FieldHasInvalidSerializeField(symbols, out var fieldName))
```
Comment (jbevain) [inline src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs]: This looks suspicious, I think we need a new test for that scenario. One FieldDeclaration can declare many field, we need to test the following scenario, which is valid:

```csharp
[SerializeField] private Transform transformA, transformB;
```

And the following scenario which is redundant:


```csharp
[SerializeField] public Transform transformA, transformB;
```

Comment (jbevain) [reply src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs]: C# doesn't let you add an attribute per field on the same declaration, so that helps.

Comment (AnBucyk) [reply src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs]: Yeah, that makes this pretty straightforward. I'll streamline this a bit and add new test cases.

File: src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+		{
+			context.EnableConcurrentExecution();
+			context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+			context.RegisterSyntaxNodeAction(AnalyzePropertyDeclaration, SyntaxKind.PropertyDeclaration);
+			context.RegisterSyntaxNodeAction(AnalyzeFieldDeclaration, SyntaxKind.FieldDeclaration);
+		}
+
+		private static void AnalyzePropertyDeclaration(SyntaxNodeAnalysisContext context)
+		{
+			var property = (MemberDeclarationSyntax)context.Node;
```
Comment (sailro) [inline src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs]: Perhaps we can use instead ?
```csharp
			var property = (PropertyDeclarationSyntax)context.Node;
			if (property == null)
				return;
```

File: src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+			if (!PropertyHasInvalidSerializeField(symbol, out var propertyName))
+				return;
+
+			context.ReportDiagnostic(Diagnostic.Create(Rule, property.GetLocation(), propertyName));
+		}
+
+		private static void AnalyzeFieldDeclaration(SyntaxNodeAnalysisContext context)
+		{
+			var field = (FieldDeclarationSyntax)context.Node;
+			if (!(field is FieldDeclarationSyntax))
```
Comment (sailro) [inline src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs]:
```
			if (field == null)
```

File: src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+				return;
+
+			List<ISymbol> symbols = new List<ISymbol>();
```
Comment (sailro) [inline src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs]:
```
var symbols = new List<ISymbol>();
```

File: src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+			foreach (var variable in field.Declaration.Variables)
+			{
+				var symbol = context.SemanticModel.GetDeclaredSymbol(variable);
+				if (symbol != null)
+					symbols.Add(symbol);
+			}
+
+			if (symbols.FirstOrDefault() == null)
```
Comment (sailro) [inline src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs]: Perhaps we can use this instead ? (to double-check)
```
if (!symbols.Any())
```

File: src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+				return false;
+
+			propertyName = symbol.Name;
+			return true;
+		}
+
+		private static bool FieldHasInvalidSerializeField(List<ISymbol> symbols, out string fields)
+		{
+			fields = null;
+			List<string> fieldNames = new List<string>();
```
Comment (sailro) [inline src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs]:
```
var fieldNames = new List<string>();
```

File: src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+	{
+		public sealed override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(ImproperSerializeFieldAnalyzer.Id);
+
+		public sealed override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;
+
+		public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+		{
+			var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+
+			var declaration = root.FindNode(context.Span) as MemberDeclarationSyntax;
```
Comment (sailro) [inline src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs]: We can use pattern matching here:
```
			if (!(root.FindNode(context.Span) is MemberDeclarationSyntax declaration))
				return;
```

File: src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+			if (declaration == null)
+				return;
+
+			if (!(declaration is PropertyDeclarationSyntax || declaration is FieldDeclarationSyntax))
+				return;
+
+			context.RegisterCodeFix(
+				CodeAction.Create(
+					Strings.ImproperSerializeFieldCodeFixTitle,
+					ct => RemoveSerializeFieldAttribute(context.Document, context, declaration, ct),
```
Comment (sailro) [inline src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs]: context is never used inside the called method
```
					ct => RemoveSerializeFieldAttribute(context.Document, declaration, ct),
```

File: src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+					declaration.ToFullString()),
+				context.Diagnostics);
+		}
+
+		private static async Task<Document> RemoveSerializeFieldAttribute(Document document, CodeFixContext context, MemberDeclarationSyntax declaration, CancellationToken cancellationToken)
```
Comment (sailro) [inline src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs]:
```
private static async Task<Document> RemoveSerializeFieldAttribute(Document document, MemberDeclarationSyntax declaration, CancellationToken cancellationToken)
```

File: src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+		{
+			var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+			var model = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+			var attributes = new SyntaxList<AttributeListSyntax>();
+
+			foreach (var attributeList in declaration.AttributeLists)
+			{
+				List<AttributeSyntax> nodes = new List<AttributeSyntax>();
```
Comment (sailro) [inline src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs]:
```
var nodes = new List<AttributeSyntax>();
```

File: src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+				foreach (var node in attributeList.Attributes)
+				{
+					var attributeType = model.GetTypeInfo(node);
+					if (attributeType.Type.Matches(typeof(UnityEngine.SerializeField)))
+						nodes.Add(node);
+				}
+
+				if (nodes.Count() > 0)
```
Comment (sailro) [inline src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs]: Perhaps we can remove the block and use to reduce nesting:
```
				if (!nodes.Any()) 
					continue;
```

**Comment by AnBucyk** [comment]
I found another bug I'm working out in the analyzer when you declare a field with multiple attributes. The code fix suggestion is a bit too aggressive, and I need to improve how I'm rebuilding the attribute list. Some testing makes me suspect I need to do so manually using the SyntaxFactory.

File: src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs (hunk 3 above)
Comment (AnBucyk) [reply src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs]: You're right. This is left over from when I was trying to analyze both Properties and Fields in the same method. Turns out to be a large difference than I thought!

File: src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs (hunk 12 above)
Comment (AnBucyk) [reply src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs]: There's now an Else here, so not exactly, but I've incorporated other parts of this feedback into an upcoming commit.

**Comment by jbevain** [comment]
Let's take Seb's suggestion and get that squashed in. Thanks!

**Pull request merged**